packages/pyright-internal/src/analyzer/parseTreeUtils.ts

```diff
@@ -96,25 +96,46 @@ export function findNodeByOffset(node: ParseNode, offset: number): ParseNode | u
 
     // The range is found within this node. See if we can localize it
     // further by checking its children.
-    const children = getChildNodes(node);
+    let children = getChildNodes(node);
```

Comment:
Rather than jumping through all of these hoops to decide whether the range rules should be followed, would it make sense to push this to the layer below and have getChildNodes do the sorting — and perhaps even filtering out of null children? We could create a variant called getSorderChildNodes or something like that. It would simplify this logic significantly.

Reply:
I can make getChildNode not to return null value, but having node in sorted (and not overlapping) is not possible. for example, this change - https://github.com/microsoft/pyright/issues/4368 - made parse tree to look like this

```python
left = middle = right = 1

[112419] 'c:\pyTemp\junk\test.py:1:1' => Module (0,0)-(4,17) =>
  [112420] 'c:\pyTemp\junk\test.py:1:1' => StatementList (0,0)-(0,25) =>
    [112427] 'c:\pyTemp\junk\test.py:1:1' => Assignment (0,0)-(0,25) =>
      [112423] 'c:\pyTemp\junk\test.py:1:17' => Name (0,16)-(0,21) => 'c:\pyTemp\junk\test.py:1:17' (Identifier, (0,16)-(0,21)) {"start":16,"length":5,"type":7,"value":"right"} right
      [112426] 'c:\pyTemp\junk\test.py:1:1' => Assignment (0,0)-(0,25) =>
        [112422] 'c:\pyTemp\junk\test.py:1:8' => Name (0,7)-(0,13) => 'c:\pyTemp\junk\test.py:1:8' (Identifier, (0,7)-(0,13)) {"start":7,"length":6,"type":7,"value":"middle"} middle
        [112425] 'c:\pyTemp\junk\test.py:1:1' => Assignment (0,0)-(0,25) =>
          [112421] 'c:\pyTemp\junk\test.py:1:1' => Name (0,0)-(0,4) => 'c:\pyTemp\junk\test.py:1:1' (Identifier, (0,0)-(0,4)) {"start":0,"length":4,"type":7,"value":"left"} left
          [112424] 'c:\pyTemp\junk\test.py:1:25' => Number (0,24)-(0,25) => 1 int:(true) imaginary:(false)
```

basically, it made tree to be evaluation order not textual order and child's span is overlapping (start is same).

we could sort by both start and end location and iterate nodes until next sibling no longer contain the given position. but it might need to create more temp arrays and more if checks.

current code is to minimize those cases.

Reply:
OK, then I think your approach is fine.